vpr/src/place/place_constraints.cpp

```diff
@@ -141,3 +141,68 @@ bool cluster_floorplanning_check(ClusterBlockId blk_id, t_pl_loc loc) {
 
     return floorplanning_good;
 }
+
+void check_cluster_constraints_init(std::string constraints_file) {
+    auto& floorplanning_ctx = g_vpr_ctx.mutable_floorplanning();
+    auto& cluster_ctx = g_vpr_ctx.clustering();
+
+    if (floorplanning_ctx.cluster_constraints.size() == 0 && constraints_file.empty()) {
+        floorplanning_ctx.cluster_constraints.resize(cluster_ctx.clb_nlist.blocks().size());
+
+        //Assign an empty PartitionRegion to each cluster block
+        for (auto blk_id : cluster_ctx.clb_nlist.blocks()) {
+            PartitionRegion empty_pr;
+            floorplanning_ctx.cluster_constraints[blk_id] = empty_pr;
+        }
+    } else if (floorplanning_ctx.cluster_constraints.size() == 0 && !constraints_file.empty()) {
+        load_cluster_constraints();
+    }
+}
+
+void load_cluster_constraints() {
+    auto& floorplanning_ctx = g_vpr_ctx.mutable_floorplanning();
+    auto& cluster_ctx = g_vpr_ctx.clustering();
+    auto& atom_ctx = g_vpr_ctx.atom();
+
+    //See which clusters contain constrained atoms
+    //Need this information to load cluster_constraints accordingly
+    vtr::vector<ClusterBlockId, std::vector<AtomBlockId>> cluster_constrained_atoms;
+
+    cluster_constrained_atoms.resize(cluster_ctx.clb_nlist.blocks().size());
+
+    for (auto atom_blk_id : atom_ctx.nlist.blocks()) {
+        PartitionId partid = floorplanning_ctx.constraints.get_atom_partition(atom_blk_id);
+
+        //If an atom returns a valid PartitionId it is part of a partition (i.e. has floorplan constraints)
+        if (partid != PartitionId::INVALID()) {
+            ClusterBlockId clb_index = atom_ctx.lookup.atom_clb(atom_blk_id);
+            cluster_constrained_atoms[clb_index].push_back(atom_blk_id);
+        }
+    }
+
+    floorplanning_ctx.cluster_constraints.resize(cluster_ctx.clb_nlist.blocks().size());
+
+    for (auto cluster_id : cluster_ctx.clb_nlist.blocks()) {
```

Comment:
refactor: for every cluster, you have an empty PR, go through the atoms of that cluster in a loop and intersect with the empty PR and update.